networkx/convert_matrix.py

```diff
@@ -1360,3 +1364,192 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
     return from_numpy_matrix(
         A, parallel_edges=parallel_edges, create_using=create_using
     )
+
+
+def generate_node_dataframe(
```

Comment:
The naming scheme for the rest of this module is to use ```to``` instead of ```generate```. 
 For example: Is ```to_node_dataframe``` a poor choice for this function?
Even if that doesn't work, I would encourage us not to use ```generate_XXX``` as we already have Graph generators and that name has been problematic because of Python's use of the word ```generator```. So I worry that this naming choice might be problematic too. 

I think we should also decide how to name functions that do this very thing but create a numpy array instead of a pandas dataframe. I believe that originally this PR had a keyword option to enable that... but it actually constructed the pandas dataframe and then converted to numpy array in that case.  Have you thought about/tried a version that constructs the numpy.array and then adds the names to the rows and columns when it converts to a dataframe? Same sort of idea when you construct a list of Series building it into the dataframe.  Would it be better to construct the dataframe directly?

I don't know when it is best to construct with lists/dicts, with numpy.arrays or with pandas dataframes or xarrays?  Do any of them present a way to input directly, or do they all come back to constructing the data as lists/dicts and then reading it into the object?

Reply:
> The naming scheme for the rest of this module is to use to instead of generate.
> For example: Is to_node_dataframe a poor choice for this function?

Not at all! I have changed the function name to use `to_`. :smile:

> Have you thought about/tried a version that constructs the numpy.array and then adds the names to the rows and columns when it converts to a dataframe? Same sort of idea when you construct a list of Series building it into the dataframe. Would it be better to construct the dataframe directly?

Syntactically, it's more compact to construct the DataFrame/Series first before converting to an array. It's more complex, I think, to construct the array and then tag on the metadata. Something about the latter makes me less at ease, in that it's easier in our codebase to mess up the labelling of the DataFrame by constructing of the array first. (After all, it's easy to get the shape right while also getting the semantic ordering/labelling of rows/columns wrong.) We externalize that level of complexity by calling on pandas' array-returning methods/attributes after-the-fact.

> I don't know when it is best to construct with lists/dicts, with numpy.arrays or with pandas dataframes or xarrays? Do any of them present a way to input directly, or do they all come back to constructing the data as lists/dicts and then reading it into the object?

The way I think about it is as follows: in a hierarchy of complexity, DataFrames are the "most complex", followed by arrays, followed by Python primitive data structures. Though they are the most complex, DataFrames also provide convenience methods to drop down to arrays and lists if a user so desires. Because it's easy for an end-user to do so, by either a `df[column].to_list()` for lists or `df.values` for arrays, I think Ross' previous comment on externalizing this piece to the end-user is a good rule of thumb to follow.